### PR TITLE
Stake: pay rewards during unbond

### PR DESCRIPTION
**Pull request by gangov**

### CODE REVIEW

File: contracts/stake/src/contract.rs [hunk 1]
```diff
@@ -280,43 +285,7 @@ impl StakingTrait for Staking {
     }
 
     fn withdraw_rewards(env: Env, sender: Address) {
-        env.events().publish(("withdraw_rewards", "user"), &sender);
-
```
Comment (ueco-jb) [inline contracts/stake/src/contract.rs:284]: Bring it back as it was.

File: contracts/stake/src/contract.rs [hunk 2]
```diff
@@ -181,6 +181,11 @@ impl StakingTrait for Staking {
 
         let config = get_config(&env);
 
+        let withdrawable_rewards = get_withdrawable_rewards(&env, &sender);
+        if !withdrawable_rewards.rewards.is_empty() {
+            do_withdraw_rewards(&env, &sender);
```
Comment (ueco-jb) [inline contracts/stake/src/contract.rs:186]:
```suggestion
            withdraw_rewards(&env, &sender);
```

**Comment by ueco-jb** [comment]
Also you did nothing on multihop, so the title is misleading.

File: contracts/stake/src/contract.rs [hunk 3]
```diff
@@ -442,24 +447,7 @@ impl StakingTrait for Staking {
...
-            let reward_amount =
-                withdrawable_rewards(&env, &user, &distribution, &withdraw_adjustment);
-            rewards.push_back(WithdrawableReward {
-                reward_address: distribution_address,
-                reward_amount,
-            });
-        }
-
-        WithdrawableRewardsResponse { rewards }
+        get_withdrawable_rewards(&env, &user)
```
Comment (ueco-jb) [inline contracts/stake/src/contract.rs:450]: Bring it the way it was, I wrote it already before.

File: contracts/stake/src/distribution.rs [hunk 4]
```diff
@@ -216,3 +219,61 @@ pub fn calculate_annualized_payout(reward_curve: Option<Curve>, now: u64) -> Dec
         None => Decimal::zero(),
     }
 }
+
+pub fn get_withdrawable_rewards(env: &Env, user: &Address) -> WithdrawableRewardsResponse {
```
Comment (ueco-jb) [inline contracts/stake/src/distribution.rs:223]: Why did you duplicate the same function from the `contract.rs`?

File: contracts/stake/src/distribution.rs [hunk 5]
```diff
@@ -216,3 +219,61 @@ pub fn calculate_annualized_payout(reward_curve: Option<Curve>, now: u64) -> Dec
...
+        rewards.push_back(WithdrawableReward {
+            reward_address: distribution_address,
+            reward_amount,
+        });
+    }
+
+    WithdrawableRewardsResponse { rewards }
+}
+
+pub fn withdraw_rewards(env: &Env, sender: &Address) {
```
Comment (ueco-jb) [inline contracts/stake/src/distribution.rs:243]: Why did you duplicate the same function from the `contract.rs`?

Comment (gangov) [reply contracts/stake/src/distribution.rs:243]: so I can reuse it in `unbond`:
> check for withdrawable rewards
> do_withdraw if some are found

File: contracts/stake/src/distribution.rs [hunk 6]
```diff
@@ -216,3 +216,41 @@ pub fn calculate_annualized_payout(reward_curve: Option<Curve>, now: u64) -> Dec
...
+        );
+
+        env.events().publish(
+            ("withdraw_rewards", "reward_token"),
+            &reward_token_client.address,
+        );
+        env.events()
+            .publish(("withdraw_rewards", "reward_amount"), reward_amount as i128);
+    }
+}
```
Comment (ueco-jb) [inline contracts/stake/src/distribution.rs:256]: To remove...

```suggestion
```

File: contracts/stake/src/tests/bond.rs [hunk 7]
```diff
@@ -226,3 +226,57 @@ fn unbond_wrong_user_stake_not_found() {
...
+
+    lp_token.mint(&user, &AMOUNT);
+    lp_token.mint(&owner, &AMOUNT);
+
+    staking.create_distribution_flow(&owner, &lp_token.address);
+    staking.fund_distribution(&owner, &0u64, &10_000u64, &lp_token.address, &AMOUNT);
+
+    env.ledger().with_mut(|li| {
+        li.timestamp = WITHDRAW_TIMESTAMP;
+    });
```
Comment (ueco-jb) [inline contracts/stake/src/tests/bond.rs:254]: You created a distribution ranging from timestamp 0 to timestamp 10_000, and you immediately jump 2000 seconds ahead. Why?

File: contracts/stake/src/tests/bond.rs [hunk 8]
```diff
@@ -226,3 +226,57 @@ fn unbond_wrong_user_stake_not_found() {
...
+            .rewards
+            .iter()
+            .map(|reward| reward.reward_amount)
+            .sum::<u128>(),
+        0
+    );
+    assert_eq!(
+        staking.query_undistributed_rewards(&lp_token.address),
+        110_000
+    );
```
Comment (ueco-jb) [inline contracts/stake/src/tests/bond.rs:274]: You have funded distribution with 100_000 tokens, why are undistributed rewards 110k?

File: contracts/stake/src/tests/bond.rs (hunk 7 above)
Comment (gangov) [reply contracts/stake/src/tests/bond.rs:254]: I chose numbers based on other places in our code where we change the ledger().timestamp. Do the numbers we use matter?

File: contracts/stake/src/tests/bond.rs (hunk 8 above)
Comment (gangov) [reply contracts/stake/src/tests/bond.rs]: thought the distribution was happening automatically, looking at the other tests, we have to call `stake.distribute_rewards`; added this missing piece to the tests and rewrote the assertions

**Review by ueco-jb — APPROVED** [review]
LGTM
Add this to the changelog please.

**Pull request merged**